### PR TITLE
Add slotscheck

### PR DESCRIPTION
**Pull request by ariebovenberg**
# I have made things!

@sobolevn kindly [invited me](https://t.me/drypython/6689) to add [slotscheck](https://github.com/ariebovenberg/slotscheck) to the CI.

## Checklist

- [x] I have double checked that there are no unrelated changes in this pull request (old patches, accidental config files, etc)
- [ ] ~~I have created at least one test case for the changes I have made~~ n/a
- [x] I have updated the documentation for the changes I have made
- [x] I have added my changes to the `CHANGELOG.md`

## Related issues

related: #1147, #1144 

## Open questions

- I've kept the same slots rules as present in `test_slots.py`. This means `returns.contrib.mypy` is not checked. We can either:
  a) add slots to `mypy` plugin classes
  b) keep the current ignore
  c) run `slotscheck` on this part of the code with more relaxed settings (there is an [issue open](https://github.com/ariebovenberg/slotscheck/issues/33) multi-setting support if you have an opinion)
- The `UnwrapFailedError` class reports an error. This is because `BaseException` actually has a `__dict__`. We can either:
  a) remove slots from `UnwrapFailedError`
  b) keep the current ignore

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -73,6 +80,7 @@ flake8-pyi = "^20.10"
 # correctly with Python 3.10. That lib is a dependency of `nitpick`!
 nitpick = { version = "^0.29", python = "<3.10" }
 codespell = "^2.1"
+slotscheck = "^0.5.1"
```
Comment (sobolevn) [inline pyproject.toml:83]:
```suggestion
slotscheck = "^0.5"
```

File: pyproject.toml [hunk 2]
```diff
@@ -6,6 +6,13 @@ build-backend = "poetry.core.masonry.api"
 [tool.nitpick]
 style = "https://raw.githubusercontent.com/wemake-services/wemake-python-styleguide/master/styles/nitpick-style-wemake.toml"
 
+[tool.slotscheck]
+strict-imports = true
+require-subclass = true
+require-superclass = true
+exclude-modules = 'returns\.contrib\.mypy'
+exclude-classes = 'returns\.primitives\.exceptions:UnwrapFailedError'
```
[the remark this answers is not in the document]
Comment (ariebovenberg) [reply pyproject.toml:14]: `BaseException` doesn't have slots

**Comment by sobolevn** [comment]
> I've kept the same slots rules as present in test_slots.py. This means returns.contrib.mypy is not checked. We can either:

Let's go with `b`. No need to check `mypy` plugin for slots.

> The UnwrapFailedError class reports an error. This is because BaseException actually has a __dict__. We can either:

I think having a `__slot__` definition is still makes attr access a bit faster, doesn't it? If so, I am happy with special ignoring it.

**Comment by ariebovenberg** [comment]
- Indeed there is still an advantage to slots, even if the base class doesn't support it. It may be common enough to add a `--no-require-exceptions` flag to `slotscheck` to disable slots check for exception subclasses.
- any idea what the docs error is about? (Before I start debugging locally)

**Comment by sobolevn** [comment]
@ariebovenberg it is unrelated, I broke it recently somehow (sphinx 😒 ).

Fix in a separate PR is highly appreciated.

**Review by sobolevn — APPROVED** [review]
Thanks again!

**Comment by ariebovenberg** [comment]
> @ariebovenberg it is unrelated, I broke it recently somehow (sphinx 😒 ).
> 
> Fix in a separate PR is highly appreciated.

An issue is a start: #1234

**Pull request merged**